src/stan/io/reader.hpp

```diff
@@ -1227,116 +1317,351 @@ class reader {
   }
 
   template <typename TL>
-  inline row_vector_t row_vector_lb_constrain(const TL lb, size_t m, T &lp) {
+  inline row_vector_t row_vector_lb_constrain(const TL lb, size_t m, T& lp) {
     row_vector_t v(m);
     for (size_t i = 0; i < m; ++i)
       v(i) = scalar_lb_constrain(lb, lp);
     return v;
   }
 
   template <typename TL>
-  inline matrix_t matrix_lb(const TL lb, const size_t m, size_t n) {
-    matrix_t v(m, n);
-    for (size_t j = 0; j < n; ++j)
-      for (size_t i = 0; i < m; ++i)
+  inline matrix_t matrix_lb(const TL lb, const size_t n, size_t m) {
+    matrix_t v(n, m);
+    for (size_t j = 0; j < m; ++j)
+      for (size_t i = 0; i < n; ++i)
         v(i, j) = scalar_lb(lb);
     return v;
   }
 
   template <typename TL>
-  inline matrix_t matrix_lb_constrain(const TL lb, size_t m, size_t n) {
-    matrix_t v(m, n);
-    for (size_t j = 0; j < n; ++j)
-      for (size_t i = 0; i < m; ++i)
+  inline matrix_t matrix_lb_constrain(const TL lb, size_t n, size_t m) {
+    matrix_t v(n, m);
+    for (size_t j = 0; j < m; ++j)
+      for (size_t i = 0; i < n; ++i)
         v(i, j) = scalar_lb_constrain(lb);
     return v;
   }
 
   template <typename TL>
-  inline matrix_t matrix_lb_constrain(const TL lb, size_t m, size_t n, T &lp) {
-    matrix_t v(m, n);
-    for (size_t j = 0; j < n; ++j)
-      for (size_t i = 0; i < m; ++i)
+  inline matrix_t matrix_lb_constrain(const TL lb, size_t n, size_t m, T& lp) {
+    matrix_t v(n, m);
+    for (size_t j = 0; j < m; ++j)
+      for (size_t i = 0; i < n; ++i)
         v(i, j) = scalar_lb_constrain(lb, lp);
     return v;
   }
 
+  /**
+   * Return a sparse matrix of the specified dimensionality and number of
+   * nonzero elements made up of the next scalars arranged in
+   * column-major order. Throws if any scalar value is less than lb.
+   *
+   * @tparam TL the type of the lower bound.
+   * @tparam VecR Vector with an integral type
+   * @tparam VecC Vector with an integral type
+   * @param lb The lower bound to check all scalar values against
+   * @param vec_r rows corresponding the nonzero entries.
+   * @param vec_c columns corresponding the nonzero entries.
+   * @param n Number of rows.
+   * @param m Number of columns.
+   * @return Eigen::SparseMatrix made up of the next scalars.
+   */
+  template <typename TL, typename VecR, typename VecC,
+            typename = require_all_std_vector_vt<std::is_integral, VecR, VecC>>
+  inline sparse_matrix_t sparse_matrix_lb(const TL lb, const VecR& vec_r,
+                                          const VecC& vec_c, size_t n,
+                                          size_t m) {
+    stan::math::check_matching_sizes("sparse_matrix_lb", "vec_r", vec_r,
+                                     "vec_c", vec_c);
+    Eigen::SparseMatrix<T> ret_mat(n, m);
+    ret_mat.reserve(vec_r.size());
+    if (m == 0 || n == 0) {
+      return ret_mat;
+    }
+    using triplet_type = Eigen::Triplet<T>;
```

Comment:
This looks cut-and-paste from the non-lb case.   Is there a way to write this once and plug in the reader like scalar_lb as another argument or would that lose efficiency?  that is, something like:

```
return construct_sparse_matrix(m, n, vec_r, vec_c, [](){ return scalar_lb(lb); });
```

An alternative that might not be much less efficient is to read the right number of scalar_lb into a vector, then use that in a call to a generic construction method for a sparse matrix from an integer vector, integer vector, and real vector).  Then this code would look like:

```
return to_sparse_matrix(m, n, vec_r, vec_c, vector_lb(lb, m * n));
```

or something like that, where you call the vector method with a lower bound that presumabl already exists somewhere.